src/components/timeline/Markers.js

```diff
@@ -27,6 +28,196 @@ export const MIN_MARKER_WIDTH = 0.3;
 
 type MarkerState = 'PRESSED' | 'HOVERED' | 'NONE';
 
+type MouseEventHandler = (SyntheticMouseEvent<HTMLCanvasElement>) => any;
+
+/**
+ * When adding properties to these props, please consider the comment above the component.
+ */
+type CanvasProps = {|
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +width: CssPixels,
+  +height: CssPixels,
+  +getMarker: MarkerIndex => Marker,
+  +markerIndexes: MarkerIndex[],
+  +hoveredItem: Marker | null,
+  +mouseDownItem: Marker | null,
+  +onMouseDown: MouseEventHandler,
+  +onMouseUp: MouseEventHandler,
+  +onMouseMove: MouseEventHandler,
+  +onMouseOut: MouseEventHandler,
+|};
+
+function _drawRoundedRect(
+  ctx: CanvasRenderingContext2D,
+  x: CssPixels,
+  y: CssPixels,
+  width: CssPixels,
+  height: CssPixels,
+  cornerSize: CssPixels
+) {
+  // Cut out c x c -sized squares in the corners.
+  const c = Math.min(width / 2, Math.min(height / 2, cornerSize));
+  const bottom = y + height;
+  ctx.fillRect(x + c, y, width - 2 * c, c);
+  ctx.fillRect(x, y + c, width, height - 2 * c);
+  ctx.fillRect(x + c, bottom - c, width - 2 * c, c);
+}
+
+/**
+ * This component controls the rendering of the canvas. Every render call through
+ * React triggers a new canvas render. Because of this, it's important to only pass
+ * in the props that are needed for the canvas draw call.
+ */
+class MarkersCanvas extends React.PureComponent<CanvasProps> {
+  _canvas: {| current: HTMLCanvasElement | null |} = React.createRef();
+  _requestedAnimationFrame: boolean = false;
+
+  _getMarkerState(marker: Marker): MarkerState {
+    const { hoveredItem, mouseDownItem } = this.props;
+
+    if (mouseDownItem !== null) {
+      if (marker === mouseDownItem && marker === hoveredItem) {
+        return 'PRESSED';
+      }
+      return 'NONE';
+    }
+    if (marker === hoveredItem) {
+      return 'HOVERED';
+    }
+    return 'NONE';
+  }
+
+  drawCanvas(c: HTMLCanvasElement) {
+    const {
+      rangeStart,
+      rangeEnd,
+      width,
+      height,
+      getMarker,
+      markerIndexes,
+    } = this.props;
+
+    if (height === 0 || width === 0) {
+      // bail out early if the size isn't known yet.
+      return;
+    }
```

Comment:
I added this compared to what we had before. The reason is that previously the `scheduleDraw` call was in `componentDidUpdate` and so wasn't happening at mount time. Moving it to `render` (to make it consistent with the other components) triggered an earlier draw, which wasn't useful because the size weren't set yet. Moreover this was making tests failing for something useless. With this simple condition the tests are still passing which makes me more confident that the behavior is the same than before.